igvc_navigation/src/trajectory_planner/smooth_control.cpp

```diff
@@ -0,0 +1,211 @@
+}
+
+void SmoothControl::getPath(const nav_msgs::PathConstPtr& path, const igvc_msgs::trajectoryPtr& trajectory_ptr,
+                            const RobotState& start_pos)
+{
+  RobotState state = start_pos;
+  std::optional<RobotState> simulation_target = target_;
+  size_t path_index = 0;
+
+  ros::Time time = path->header.stamp;
+  trajectory_ptr->header.stamp = path->header.stamp;
+
+  // Calculate timesteps
+  static ros::Duration dt = ros::Duration(1 / path_generation_options_.simulation_frequency);
+
+  // Visualization
+  nav_msgs::Path closest_point_path;
+  nav_msgs::Path targets_path;
+
+  closest_point_path.header = path->header;
+  targets_path.header = path->header;
+  for (int i = 0; i < path_generation_options_.getNumSamples() && path_index < path->poses.size() - 1; i++, time += dt)
+  {
+    // Get target
+    path_index = getClosestIndex(path, state);
+    RobotState second_target;
+    std::tie(simulation_target, second_target) = getTargetPosition(path, state, simulation_target, path_index);
+
+    RobotControl control = getControl(state, *simulation_target, second_target);
+    state.propogateState(control, dt.toSec());
+
+    geometry_msgs::PoseStamped pose_stamped;
+    pose_stamped.header = path->header;
+    pose_stamped.pose = path->poses[path_index].pose;
+    closest_point_path.poses.emplace_back(pose_stamped);
+
+    geometry_msgs::PoseStamped target{};
+    target.pose = simulation_target->toPose();
+    target.header.stamp = time;
+    targets_path.poses.emplace_back(target);
+
+    trajectory_ptr->trajectory.emplace_back(state.toTrajectoryPoint(time, control, axle_length_));
+  }
+  target_pub_.publish(targets_path);
+  closest_point_pub_.publish(closest_point_path);
+}
+
+RobotControl SmoothControl::getControl(const RobotState& state, const RobotState& target,
+                                       const RobotState& second_target) const
+{
+  double K_first = getCurvature(state, target);
+  double K_second = getCurvature(state, second_target);
+  double distance = state.distTo(target);
+  double K = blendCurvature(distance, K_first, K_second);
+  return RobotControl::fromKV(K, path_generation_options_.simulation_velocity, axle_length_);
+}
+
+double SmoothControl::getCurvature(const RobotState& state, const RobotState& target) const
+{
+  double line_of_sight = atan2(target.y() - state.y(), target.x() - state.x());
+
+  double delta = state.yaw() - line_of_sight;
+  double theta = target.yaw() - line_of_sight;
+
+  // adjust both angles to lie between -PI and PI
+  igvc::fit_to_polar(delta);
+  igvc::fit_to_polar(theta);
+
+  // calculate the radius of curvature, K
+  double d = state.distTo(target);  // euclidian distance to target
+  double K = smooth_control_options_.k2 * (delta - atan(-smooth_control_options_.k1 * theta));
+  K += (1 + (smooth_control_options_.k1 / (1 + pow(smooth_control_options_.k1 * theta, 2)))) * sin(delta);
+  K /= -d;
```

Comment:
A few notes:
 - It doesn't make sense to mutate `K` here, the intermediaries have nothing to do with the curvature.
 - Use some actual intermediary equations here. Check the original paper (also probably cite it here) to see if they have any.
 - Cut down on visual noise: `const auto [k1, k2] = smooth_control_options_` already does a ton to clean this up, for example.

Reply:
I agree that it doesn't make sense to mutate `K`, but like the [paper](https://web.eecs.umich.edu/~kuipers/papers/Park-icra-11.pdf) doesn't really have that many intermediary equations.